pkg/component/controller/konnectivityagent.go

```diff
@@ -0,0 +1,301 @@
+func (k *KonnectivityAgent) Reconcile(ctx context.Context, clusterCfg *v1beta1.ClusterConfig) error {
+	k.configChangeChan <- clusterCfg
+
+	return nil
+}
+
+func (k *KonnectivityAgent) Stop() error {
+	return nil
+}
+
+func (k *KonnectivityAgent) writeKonnectivityAgent() error {
+	k.agentManifestLock.Lock()
+	defer k.agentManifestLock.Unlock()
+
+	if k.clusterConfig == nil {
+		return fmt.Errorf("cluster config is not reconciled yet")
+	}
+
+	konnectivityDir := filepath.Join(k.K0sVars.ManifestsDir, "konnectivity")
+	err := dir.Init(konnectivityDir, constant.ManifestsDirMode)
+	if err != nil {
+		return err
+	}
+	cfg := konnectivityAgentConfig{
+		// Since the konnectivity server runs with hostNetwork=true this is the
+		// IP address of the master machine
+		ProxyServerHost: k.NodeConfig.Spec.API.APIAddress(), // TODO: should it be an APIAddress?
+		ProxyServerPort: uint16(k.clusterConfig.Spec.Konnectivity.AgentPort),
+		Image:           k.clusterConfig.Spec.Images.Konnectivity.URI(),
+		ServerCount:     k.serverCount,
+		PullPolicy:      k.clusterConfig.Spec.Images.DefaultPullPolicy,
+	}
+
+	if k.clusterConfig.Spec.Network != nil {
+		nllb := k.clusterConfig.Spec.Network.NodeLocalLoadBalancing
+		if nllb.IsEnabled() {
+			switch nllb.Type {
+			case v1beta1.NllbTypeEnvoyProxy:
+				k.log.Debugf("Enabling node-local load balancing via %s", nllb.Type)
+
+				// FIXME: Transitions from non-node-local load balanced to
+				// node-local load balanced setups will be problematic: The
+				// controller will update the DaemonSet with localhost, but the
+				// worker nodes won't reconcile their state (yet) and need to be
+				// restarted manually in order to start their load balancer.
+				// Transitions in the other direction suffer from the same
+				// limitation, but that will be less grave, as the node-local
+				// load balancers will remain operational until the next node
+				// restart and the agents will stay connected.
+
+				// The node-local load balancer will run in the host network, so
+				// the agent needs to do the same in order to use it.
+				cfg.HostNetwork = true
+
+				// FIXME: This is not exactly on par with the way it's
+				// implemented on the worker side, i.e. there's no fallback if
+				// localhost doesn't resolve to a loopback address. But this
+				// would require some shenanigans to pull in node-specific
+				// values here. A possible solution would be to convert the
+				// konnectivity agent to a static Pod as well.
+				cfg.ProxyServerHost = "localhost"
+
+				if nllb.EnvoyProxy.KonnectivityServerBindPort != nil {
+					cfg.ProxyServerPort = uint16(*nllb.EnvoyProxy.KonnectivityServerBindPort)
+				} else {
+					cfg.ProxyServerPort = uint16(*v1beta1.DefaultEnvoyProxy().KonnectivityServerBindPort)
+				}
+			default:
+				return fmt.Errorf("unsupported node-local load balancer type: %q", k.clusterConfig.Spec.Network.NodeLocalLoadBalancing.Type)
+			}
+		}
+	}
+
+	if cfg == k.previousConfig {
```

Comment:
This only works when `cfg` or `k.previousConfig` is nil. I'm not sure why the compiler doesn't catch this as it will crash when that is not the case.

```suggestion
	if cfg.Equals(k.previousConfig) {
```

Similar looking mistake seems to exist in several components 🤔

Reply:
The compiler should catch it 🤔 https://go.dev/play/p/D_lg51bSbZH

Reply:
Ah it isn't stringmap, got confused somehow.

Structs should be comparable if they only include Comparable fields.